### PR TITLE
BasicConstraint CSR and Singing

### PR DESCRIPTION
**Pull request by tam7t**
This PR resolves https://github.com/cloudflare/cfssl/issues/495 by:
- Includes `BasicConstraint` extension when generating CA singing requests
- Parses `BasicConstraint` from CSRs when signing requests

### CODE REVIEW

File: csr/csr_test.go [hunk 1]
```diff
@@ -117,6 +118,128 @@ func TestParseRequest(t *testing.T) {
...
+	csr, err := x509.ParseCertificateRequest(block.Bytes)
+	if err != nil {
+		t.Fatalf("%v", err)
+	}
+
+	found := false
+	for _, ext := range csr.Extensions {
+		if ext.Id.Equal(asn1.ObjectIdentifier{2, 5, 29, 19}) {
+			found = true
+			// TODO: make sure CA is true and pathlen is -1
```
Comment (lziest) [inline csr/csr_test.go]: I happen to feel that is important. :)

**Comment by lziest** [comment]
generally LGTM

File: csr/csr_test.go (hunk 1 above)
Comment (tam7t) [inline csr/csr_test.go]: Ah yes, I'll fix that!

**Comment by lziest** [comment]
LGTM

**Comment by lziest** [comment]
Now I could be singing. :)

**Pull request merged**